### PR TITLE
Add set `insert_and_find` device ref APIs

### PR DESCRIPTION
**Pull request by PointKernel**
This PR adds device side `insert_and_find` APIs for `static_set`. It also includes small cleanups of the current refactoring code.

### CODE REVIEW

File: include/cuco/detail/static_set/static_set_ref.inl [hunk 1]
```diff
@@ -187,68 +246,137 @@ class operator_impl<op::insert_tag,
...
+      auto const [state, intra_window_index] = [&]() {
+        for (auto i = 0; i < window_size; ++i) {
+          switch (ref_.predicate_(window_slots[i], value)) {
+            case detail::equal_result::EMPTY: return cuco::pair{detail::equal_result::EMPTY, i};
+            case detail::equal_result::EQUAL: return cuco::pair{detail::equal_result::EQUAL, i};
+            default: continue;
+          }
         }
+        // returns dummy index `-1` for UNEQUAL
+        return cuco::pair<detail::equal_result, int32_t>{detail::equal_result::UNEQUAL, -1};
```
Comment (PointKernel) [inline include/cuco/detail/static_set/static_set_ref.inl:347]: we could use a struct to get rid of `cuco::pair` here

File: include/cuco/detail/static_set/static_set_ref.inl [hunk 2]
```diff
@@ -82,6 +83,61 @@ static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::e
...
+  static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::insert_result
+  static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::attempt_insert(
+    value_type* slot, value_type const& value)
+{
+  // temporary workaround due to performance regression
+  // https://github.com/NVIDIA/libcudacxx/issues/366
+  value_type const old = [&]() {
+    value_type expected = this->empty_key_sentinel();
+    value_type val      = value;
+    if constexpr (sizeof(value_type) == sizeof(uint32_t)) {
```
Comment (sleeepyjack) [inline include/cuco/detail/static_set/static_set_ref.inl]: Nitpicking:
```suggestion
    if constexpr (sizeof(value_type) == sizeof(unsigned int)) {
```

File: include/cuco/detail/static_set/static_set_ref.inl [hunk 3]
```diff
@@ -82,6 +83,61 @@ static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::e
...
+        return atomicCAS_system(reinterpret_cast<unsigned int*>(slot), *expected_ptr, *value_ptr);
+      } else if constexpr (Scope == cuda::thread_scope_device) {
+        return atomicCAS(reinterpret_cast<unsigned int*>(slot), *expected_ptr, *value_ptr);
+      } else if constexpr (Scope == cuda::thread_scope_block) {
+        return atomicCAS_block(reinterpret_cast<unsigned int*>(slot), *expected_ptr, *value_ptr);
+      } else {
+        static_assert(cuco::dependent_false<decltype(Scope)>, "Unsupported thread scope");
+      }
+    }
+    if constexpr (sizeof(value_type) == sizeof(uint64_t)) {
```
Comment (sleeepyjack) [inline include/cuco/detail/static_set/static_set_ref.inl]:
```suggestion
    if constexpr (sizeof(value_type) == sizeof(unsigned long long int)) {
```

File: include/cuco/detail/static_set/static_set_ref.inl [hunk 4]
```diff
@@ -82,6 +83,61 @@ static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>::e
...
+          reinterpret_cast<unsigned long long int*>(slot), *expected_ptr, *value_ptr);
+      } else {
+        static_assert(cuco::dependent_false<decltype(Scope)>, "Unsupported thread scope");
+      }
+    }
+  }();
+  if (*slot == old) {
+    // Shouldn't use `predicate_` operator directly since it includes a redundant bitwise compare
+    return predicate_.equal_to(old, value) == detail::equal_result::EQUAL ? insert_result::DUPLICATE
+                                                                          : insert_result::CONTINUE;
```
Comment (sleeepyjack) [inline include/cuco/detail/static_set/static_set_ref.inl:135]: Is this a TODO/FIXME?

File: include/cuco/detail/static_set/static_set_ref.inl [hunk 5]
```diff
@@ -187,68 +246,136 @@ class operator_impl<op::insert_tag,
...
-   * @param slot Pointer to the slot in memory
-   * @param value Element to insert
+   * @param value The element to insert
    *
-   * @return Result of this operation, i.e., success/continue/duplicate
+   * @return a pair consisting of an iterator to the element and a bool indicating whether the
+   * insertion is successful or not.
    */
-  [[nodiscard]] __device__ insert_result attempt_insert(value_type* slot, value_type const& value)
+  __device__ thrust::pair<iterator, bool> insert_and_find(value_type const& value) noexcept
```
Comment (sleeepyjack) [inline include/cuco/detail/static_set/static_set_ref.inl:282]: Hmm, I'm unsure if we should return a `thrust::pair` or `cuco::pair` here.

File: include/cuco/detail/static_set/static_set_ref.inl (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/static_set/static_set_ref.inl:347]: I think we can leave it as is. No need to define a struct to make a single line of code/implementation detail look nicer.

**Review by sleeepyjack — APPROVED** [review]

File: include/cuco/detail/static_set/static_set_ref.inl (hunk 4 above)
Comment (PointKernel) [reply include/cuco/detail/static_set/static_set_ref.inl:135]: Neither `TODO` nor `FIXME`, just explaining why we cannot invoke `predicate_(old, value)` here.

File: include/cuco/detail/static_set/static_set_ref.inl (hunk 5 above)
Comment (PointKernel) [reply include/cuco/detail/static_set/static_set_ref.inl:282]: I was following the map `insert_and_find` implementation to use `thrust::pair`. `cuco::pair` requires some cleanup work like https://github.com/NVIDIA/cuCollections/pull/87 thus not mature enough to expose it.

**Pull request merged**